File: cognitive_folio/dashboard/financial_metrics/financial_metrics_dashboard.js
// Financial Metrics Dashboard

frappe.pages['financial-metrics-dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Financial Metrics Dashboard',
		single_column: true
	});

	// Add filters
	page.add_inner_button(__('Refresh'), () => load_dashboard());

	let security_filter = page.add_field({
		label: 'Security',
		fieldtype: 'Link',
		options: 'CF Security',
		fieldname: 'security',
		reqd: 1,
		change: () => load_dashboard()
	});

	let period_filter = page.add_field({
		label: 'Period Type',
		fieldtype: 'Select',
		options: 'Annual\nQuarterly',
		fieldname: 'period_type',
		default: 'Annual',
		change: () => load_dashboard()
	});

	let years_filter = page.add_field({
		label: 'Years',
		fieldtype: 'Select',
		options: '1\n3\n5\nAll',
		fieldname: 'years',
		default: '5',
		change: () => load_dashboard()
	});

	// Dashboard container
	let dashboard_html = `
		<div class="dashboard-container">
			<div class="row">
				<div class="col-md-6">
					<div class="chart-container" id="revenue-trend-chart"></div>
				</div>
				<div class="col-md-6">
					<div class="chart-container" id="margin-trends-chart"></div>
				</div>
			</div>
			<div class="row">
				<div class="col-md-6">
					<div class="chart-container" id="cash-flow-trend-chart"></div>
				</div>
				<div class="col-md-6">
					<div class="chart-container" id="quarterly-comparison-chart"></div>
				</div>
			</div>
			<div class="row">
				<div class="col-md-6">
					<div class="chart-container" id="yoy-growth-chart"></div>
				</div>
				<div class="col-md-6">
					<div class="ratios-container" id="ratios-gauges"></div>
				</div>
			</div>
		</div>
	`;

	page.main.html(dashboard_html);

	function load_dashboard() {
		let security = security_filter.get_value();
		let period_type = period_filter.get_value();
		let years = years_filter.get_value();

		if (!security) return;

		// Load all charts
		load_revenue_trend(security, period_type, years);
		load_margin_trends(security, period_type, years);
		load_cash_flow_trend(security, period_type, years);
		load_quarterly_comparison(security, period_type);
		load_yoy_growth(security, period_type, years);
		load_ratios_gauges(security, period_type);
	}

	function load_revenue_trend(security, period_type, years) {
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_revenue_trend',
			args: { security, period_type, years },
			callback: (r) => {
				if (r.message) {
					new frappe.Chart("#revenue-trend-chart", {
						title: "Revenue Trend",
						data: r.message,
						type: 'line',
						height: 300
					});
				}
			}
		});
	}

	function load_margin_trends(security, period_type, years) {
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_margin_trends',
			args: { security, period_type, years },
			callback: (r) => {
				if (r.message) {
					new frappe.Chart("#margin-trends-chart", {
						title: "Margin Trends (%)",
						data: r.message,
						type: 'line',
						height: 300
					});
				}
			}
		});
	}

	function load_cash_flow_trend(security, period_type, years) {
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_cash_flow_trend',
			args: { security, period_type, years },
			callback: (r) => {
				if (r.message) {
					new frappe.Chart("#cash-flow-trend-chart", {
						title: "Cash Flow Trend",
						data: r.message,
						type: 'line',
						height: 300
					});
				}
			}
		});
	}

	function load_quarterly_comparison(security, period_type) {
		let current_year = new Date().getFullYear();
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_quarterly_comparison',
			args: { security, fiscal_year: current_year },
			callback: (r) => {
				if (r.message) {
					new frappe.Chart("#quarterly-comparison-chart", {
						title: `Quarterly Comparison ${current_year}`,
						data: r.message,
						type: 'bar',
						height: 300
					});
				}
			}
		});
	}

	function load_yoy_growth(security, period_type, years) {
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_yoy_growth_rates',
			args: { security, period_type, years },
			callback: (r) => {
				if (r.message) {
					new frappe.Chart("#yoy-growth-chart", {
						title: "YoY Growth Rates (%)",
						data: r.message,
						type: 'bar',
						height: 300
					});
				}
			}
		});
	}

	function load_ratios_gauges(security, period_type) {
		frappe.call({
			method: 'cognitive_folio.cognitive_folio.dashboard.financial_metrics.financial_metrics_dashboard.get_financial_ratios',
			args: { security, period_type },
			callback: (r) => {
				if (r.message) {
					let ratios = r.message;
					let html = '<h4>Financial Ratios</h4>';
					Object.keys(ratios).forEach(ratio => {
						let value = ratios[ratio];
						let color = get_ratio_color(ratio, value);
						html += `
							<div class="ratio-gauge">
								<div class="ratio-label">${ratio}</div>
								<div class="ratio-value" style="color: ${color}">${value.toFixed(2)}</div>
							</div>
						`;
					});
					$('#ratios-gauges').html(html);
				}
			}
		});
	}

	function get_ratio_color(ratio, value) {
		// Simple color coding for ratios
		if (ratio.includes('ROE') || ratio.includes('ROA')) {
			return value > 10 ? 'green' : value > 5 ? 'orange' : 'red';
		}
		if (ratio.includes('Current Ratio')) {
			return value > 1.5 ? 'green' : value > 1 ? 'orange' : 'red';
		}
		if (ratio.includes('Debt/Equity')) {
			return value < 1 ? 'green' : value < 2 ? 'orange' : 'red';
		}
		return 'black';
	}

	// Initial load if security is set
	if (security_filter.get_value()) {
		load_dashboard();
	}
};

File: cognitive_folio/dashboard/financial_metrics/financial_metrics_dashboard.py
import frappe
from frappe import _
from frappe.utils import cint, flt
import json


@frappe.whitelist()
def get_revenue_trend(security=None, period_type="Annual", years=5):
	"""Get revenue trend data for chart"""
	if not security:
		return {"labels": [], "datasets": []}

	periods = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": period_type
		},
		fields=["fiscal_year", "fiscal_quarter", "total_revenue"],
		order_by="fiscal_year DESC, fiscal_quarter DESC",
		limit=cint(years) * 4  # Allow for quarterly data
	)

	# Group and sort chronologically
	data_points = {}
	for p in periods:
		key = f"{p.fiscal_year}{' Q' + str(p.fiscal_quarter) if p.fiscal_quarter else ''}"
		data_points[key] = p.total_revenue

	labels = sorted(data_points.keys())
	values = [data_points[label] for label in labels]

	return {
		"labels": labels,
		"datasets": [{
			"name": "Revenue",
			"values": values
		}]
	}


@frappe.whitelist()
def get_margin_trends(security=None, period_type="Annual", years=5):
	"""Get margin trends data for chart"""
	if not security:
		return {"labels": [], "datasets": []}

	periods = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": period_type
		},
		fields=["fiscal_year", "fiscal_quarter", "gross_margin", "operating_margin", "net_margin"],
		order_by="fiscal_year DESC, fiscal_quarter DESC",
		limit=cint(years) * 4
	)

	data_points = {}
	for p in periods:
		key = f"{p.fiscal_year}{' Q' + str(p.fiscal_quarter) if p.fiscal_quarter else ''}"
		data_points[key] = {
			"gross": p.gross_margin or 0,
			"operating": p.operating_margin or 0,
			"net": p.net_margin or 0
		}

	labels = sorted(data_points.keys())
	gross_values = [data_points[label]["gross"] for label in labels]
	operating_values = [data_points[label]["operating"] for label in labels]
	net_values = [data_points[label]["net"] for label in labels]

	return {
		"labels": labels,
		"datasets": [
			{"name": "Gross Margin %", "values": gross_values},
			{"name": "Operating Margin %", "values": operating_values},
			{"name": "Net Margin %", "values": net_values}
		]
	}


@frappe.whitelist()
def get_cash_flow_trend(security=None, period_type="Annual", years=5):
	"""Get cash flow trend data for chart"""
	if not security:
		return {"labels": [], "datasets": []}

	periods = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": period_type
		},
		fields=["fiscal_year", "fiscal_quarter", "operating_cash_flow", "free_cash_flow"],
		order_by="fiscal_year DESC, fiscal_quarter DESC",
		limit=cint(years) * 4
	)

	data_points = {}
	for p in periods:
		key = f"{p.fiscal_year}{' Q' + str(p.fiscal_quarter) if p.fiscal_quarter else ''}"
		data_points[key] = {
			"operating": p.operating_cash_flow or 0,
			"free": p.free_cash_flow or 0
		}

	labels = sorted(data_points.keys())
	operating_values = [data_points[label]["operating"] for label in labels]
	free_values = [data_points[label]["free"] for label in labels]

	return {
		"labels": labels,
		"datasets": [
			{"name": "Operating Cash Flow", "values": operating_values},
			{"name": "Free Cash Flow", "values": free_values}
		]
	}


@frappe.whitelist()
def get_quarterly_comparison(security=None, fiscal_year=None):
	"""Get quarterly comparison data for current year"""
	if not security or not fiscal_year:
		return {"labels": [], "datasets": []}

	periods = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": "Quarterly",
			"fiscal_year": cint(fiscal_year)
		},
		fields=["fiscal_quarter", "total_revenue", "net_income"],
		order_by="fiscal_quarter"
	)

	labels = []
	revenue_values = []
	net_income_values = []

	for p in periods:
		labels.append(f"Q{p.fiscal_quarter}")
		revenue_values.append(p.total_revenue or 0)
		net_income_values.append(p.net_income or 0)

	return {
		"labels": labels,
		"datasets": [
			{"name": "Revenue", "values": revenue_values},
			{"name": "Net Income", "values": net_income_values}
		]
	}


@frappe.whitelist()
def get_yoy_growth_rates(security=None, period_type="Annual", years=5):
	"""Get YoY growth rates for chart"""
	if not security:
		return {"labels": [], "datasets": []}

	periods = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": period_type
		},
		fields=["fiscal_year", "fiscal_quarter", "revenue_growth_yoy", "net_income_growth_yoy", "eps_growth_yoy"],
		order_by="fiscal_year DESC, fiscal_quarter DESC",
		limit=cint(years) * 4
	)

	data_points = {}
	for p in periods:
		key = f"{p.fiscal_year}{' Q' + str(p.fiscal_quarter) if p.fiscal_quarter else ''}"
		data_points[key] = {
			"revenue": (p.revenue_growth_yoy or 0) * 100,
			"net_income": (p.net_income_growth_yoy or 0) * 100,
			"eps": (p.eps_growth_yoy or 0) * 100
		}

	labels = sorted(data_points.keys())
	revenue_growth = [data_points[label]["revenue"] for label in labels]
	net_income_growth = [data_points[label]["net_income"] for label in labels]
	eps_growth = [data_points[label]["eps"] for label in labels]

	return {
		"labels": labels,
		"datasets": [
			{"name": "Revenue Growth %", "values": revenue_growth},
			{"name": "Net Income Growth %", "values": net_income_growth},
			{"name": "EPS Growth %", "values": eps_growth}
		]
	}


@frappe.whitelist()
def get_financial_ratios(security=None, period_type="Annual"):
	"""Get current financial ratios for gauge charts"""
	if not security:
		return {}

	latest_period = frappe.get_all(
		"CF Financial Period",
		filters={
			"security": security,
			"period_type": period_type
		},
		fields=["roe", "roa", "current_ratio", "debt_to_equity"],
		order_by="fiscal_year DESC, fiscal_quarter DESC",
		limit=1
	)

	if not latest_period:
		return {}

	period = latest_period[0]
	return {
		"ROE": period.roe or 0,
		"ROA": period.roa or 0,
		"Current Ratio": period.current_ratio or 0,
		"Debt/Equity": period.debt_to_equity or 0
	}